network/onchain-discovery/src/client.rs

```diff
@@ -246,38 +237,39 @@ where
             trusted_version,
         );
         let req_msg = QueryDiscoverySetRequest {
-            client_known_version: trusted_version,
-            client_known_seq_num: self.latest_event_seq_num,
+            known_version: trusted_version,
         };
         let self_peer_id = self.peer_id;
-        storage_query_discovery_set(Arc::clone(&self.libra_db), req_msg)
-            .map(move |res| res.map(move |(req_msg, res_msg)| (self_peer_id, req_msg, res_msg)))
+        let libra_db = Arc::clone(&self.libra_db);
+        let f_query_storage =
+            task::spawn_blocking(move || storage_query_discovery_set(libra_db, req_msg));
+        f_query_storage.map(move |res| {
```

Comment:
forgive the naïveté, what if we just did the query directly? What do we benefit by moving it to a task that we ultimately block on?

Reply:
right, so async is a cooperative multi-tasking model and there's no preemption. if this blocking storage call takes a long time (maybe libradb -> rocksdb needs to read from disk. it needs to build a large accumulator proof /w lots of CPU intensive hashing) it will starve the tokio scheduler and our other async tasks (on the network runtime at least) since the task will not "yield" until the whole query is complete. if we use `task::spawn_blocking`, then tokio will spawn this task on a separate thread pool specifically for blocking tasks.

@lightmark says this call could be on the order of 1-10ms for clients calling `get_state_proof` from far behind, which is a relative eternity to starve the scheduler and other networking tasks. especially if we get someone spamming us with these calls...